wagtail/admin/static_src/wagtailadmin/js/vendor/urlify.js

```diff
@@ -1,182 +0,0 @@
-        'Ą': 'A', 'Č': 'C', 'Ę': 'E', 'Ė': 'E', 'Į': 'I', 'Š': 'S', 'Ų': 'U',
-        'Ū': 'U', 'Ž': 'Z'
-    };
-    var SERBIAN_MAP = {
-        'ђ': 'dj', 'ј': 'j', 'љ': 'lj', 'њ': 'nj', 'ћ': 'c', 'џ': 'dz',
-        'đ': 'dj', 'Ђ': 'Dj', 'Ј': 'j', 'Љ': 'Lj', 'Њ': 'Nj', 'Ћ': 'C',
-        'Џ': 'Dz', 'Đ': 'Dj'
-    };
-    var AZERBAIJANI_MAP = {
-        'ç': 'c', 'ə': 'e', 'ğ': 'g', 'ı': 'i', 'ö': 'o', 'ş': 's', 'ü': 'u',
-        'Ç': 'C', 'Ə': 'E', 'Ğ': 'G', 'İ': 'I', 'Ö': 'O', 'Ş': 'S', 'Ü': 'U'
-    };
-    var GEORGIAN_MAP = {
-        'ა': 'a', 'ბ': 'b', 'გ': 'g', 'დ': 'd', 'ე': 'e', 'ვ': 'v', 'ზ': 'z',
-        'თ': 't', 'ი': 'i', 'კ': 'k', 'ლ': 'l', 'მ': 'm', 'ნ': 'n', 'ო': 'o',
-        'პ': 'p', 'ჟ': 'j', 'რ': 'r', 'ს': 's', 'ტ': 't', 'უ': 'u', 'ფ': 'f',
-        'ქ': 'q', 'ღ': 'g', 'ყ': 'y', 'შ': 'sh', 'ჩ': 'ch', 'ც': 'c', 'ძ': 'dz',
-        'წ': 'w', 'ჭ': 'ch', 'ხ': 'x', 'ჯ': 'j', 'ჰ': 'h'
-    };
-
-    var ALL_DOWNCODE_MAPS = [
-        LATIN_MAP,
-        LATIN_SYMBOLS_MAP,
-        GREEK_MAP,
-        TURKISH_MAP,
-        ROMANIAN_MAP,
-        RUSSIAN_MAP,
-        UKRAINIAN_MAP,
-        CZECH_MAP,
-        SLOVAK_MAP,
-        POLISH_MAP,
-        LATVIAN_MAP,
-        ARABIC_MAP,
-        LITHUANIAN_MAP,
-        SERBIAN_MAP,
-        AZERBAIJANI_MAP,
-        GEORGIAN_MAP
-    ];
-
-    var Downcoder = {
-        'Initialize': function() {
-            if (Downcoder.map) { // already made
-                return;
-            }
-            Downcoder.map = {};
-            Downcoder.chars = [];
-            for (var i = 0; i < ALL_DOWNCODE_MAPS.length; i++) {
-                var lookup = ALL_DOWNCODE_MAPS[i];
-                for (var c in lookup) {
-                    if (lookup.hasOwnProperty(c)) {
-                        Downcoder.map[c] = lookup[c];
-                    }
-                }
-            }
-            for (var k in Downcoder.map) {
-                if (Downcoder.map.hasOwnProperty(k)) {
-                    Downcoder.chars.push(k);
-                }
-            }
-            Downcoder.regex = new RegExp(Downcoder.chars.join('|'), 'g');
-        }
-    };
-
-    function downcode(slug) {
-        Downcoder.Initialize();
-        return slug.replace(Downcoder.regex, function(m) {
-            return Downcoder.map[m];
-        });
-    }
-
-
-    function URLify(s, num_chars, allowUnicode) {
-        // changes, e.g., "Petty theft" to "petty-theft"
-        if (!allowUnicode) {
-            s = downcode(s);
-        }
-        s = s.toLowerCase(); // convert to lowercase
-        // if downcode doesn't hit, the char will be stripped here
-        if (allowUnicode) {
-            // Keep Unicode letters including both lowercase and uppercase
-            // characters, whitespace, and dash; remove other characters.
-            s = XRegExp.replace(s, XRegExp('[^-_\\p{L}\\p{N}\\s]', 'g'), '');
-        } else {
-            s = s.replace(/[^-\w\s]/g, ''); // remove unneeded chars
```

Comment:
How about instead of replacing this with the npm package, we only remove the `XRegExp` usage and turn this module into a util that exports urlify as a function, to be used by us via imports? Then, we expose something like `window.URLify = urlify` in `core.js` for the deprecation process? Therefore, we can still remove this file from being included as a JS vendor file in `_editor_js.html`.

I'm not strongly opposed to the NPM package, but it seems there have been some minor updates to the function in Django since the package was last released: https://github.com/django/django/commits/main/django/contrib/admin/static/admin/js/urlify.js

From a quick look they seem to be mostly cosmetic changes, but I also noticed that the package is not compatible with strict mode (see https://github.com/fyalavuz/node-parameterize/pull/11). Not sure how that affects us, but I think it's nicer to have complete control over the code in case we ever need to make changes in the future (e.g. adding more character maps).

Reply:
As per comments below, I did look at this approach also and decided against it. I'll take another look though. 

I also manually compared the npm package with Django's implementation (after formatting both) and they are the same. Only difference is in the custom implementation of XRegExp.

However, I get that this may be seen as too risky to use an an external dependency.

Reply:
OK, no worries. I did compare them and from a quick look they seem very similar, but not exactly the same. For example, Django and us do

https://github.com/wagtail/wagtail/blob/e425244fd8a225dda2ca840d8dd3fa0b03d95243/wagtail/admin/static_src/wagtailadmin/js/vendor/urlify.js#L173

but the `parameterize` package does:

https://github.com/fyalavuz/node-parameterize/blob/7e102260ba5d107ac5d615c88e780c5eedcccf9c/parameterize.js#L191

```js
    s = s.replace(/[^-\w\s\u4E00-\u9FA5]/g, '');  // remove unneeded chars
```

The Django/our implementation trims any trailing hyphens:

https://github.com/wagtail/wagtail/blob/e425244fd8a225dda2ca840d8dd3fa0b03d95243/wagtail/admin/static_src/wagtailadmin/js/vendor/urlify.js#L178

and the `parameterize` package does not:

<img width="245" alt="image" src="https://github.com/wagtail/wagtail/assets/6379424/ffb9efe4-1ec7-4c8f-878f-613c2a53e00f">

Also, the Django implementation does `.toLowerCase()` before doing any replacements. This was done as of https://github.com/django/django/pull/12237, and I can see that our `allowUnicode` handling of `window.URLify` fallback does it this way too. Not sure what's happening but it seems the parameterize package still handles the İstanbul case correctly anyway, which is good, but it might be something to watch out for.

If you think all these differences are OK (maybe with a few tweaks e.g. trimming trailing hyphens), then I guess we could go ahead with the package.

Reply:
Thanks for the investigation, I'll give this another crack next week. I wish Django extracted this stuff to an NPM package, would be much easier.

Maybe we just make our own package?

We have the `@wagtail/` namespace. 

I guess at that point we can just copy it all into the code anyway.